File: app/src/main/java/net/albertogarrido/studydrivetest/MainActivity.java
package net.albertogarrido.studydrivetest;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static net.albertogarrido.studydrivetest.WorkerService.CONSUMER;
import static net.albertogarrido.studydrivetest.WorkerService.PRODUCER;
import static net.albertogarrido.studydrivetest.WorkerService.WORKER;
import static net.albertogarrido.studydrivetest.WorkerService.WORKER_TYPE;

public class MainActivity extends AppCompatActivity {

    private int consumers = 0;
    private int producers = 0;
    private ItemsAdapter adapter = new ItemsAdapter();
    private List<Item> itemsList = new ArrayList<>(10);

    private Messenger producerMessenger;
    private WorkerBroadcastReceiver broadcastReceiver;
    private boolean isServiceBound;

    private ServiceConnection producerConnection = new ServiceConnection() {
        public void onServiceConnected(ComponentName className, IBinder service) {
            producerMessenger = new Messenger(service);
            isServiceBound = true;
        }

        public void onServiceDisconnected(ComponentName className) {
            producerMessenger = null;
            isServiceBound = false;
        }
    };

    @BindView(R.id.consumerCount)
    TextView consumerCount;
    @BindView(R.id.producerCount)
    TextView producerCount;
    @BindView(R.id.listItemsCount)
    TextView listItemsCount;
    @BindView(R.id.list)
    RecyclerView list;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        configureList();
        refresh();
    }

    @Override
    protected void onStart() {
        super.onStart();
        bindService(WorkerService.createIntent(this), producerConnection, Context.BIND_AUTO_CREATE);
    }

    @Override
    public void onResume() {
        super.onResume();
        broadcastReceiver = new WorkerBroadcastReceiver();
        final IntentFilter intentFilter = new IntentFilter(WORKER);
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, intentFilter);
    }

    @Override
    public void onPause() {
        super.onPause();
        if (broadcastReceiver != null)
            LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver);
        broadcastReceiver = null;
    }

    @OnClick(R.id.addConsumer)
    public void onAddConsumerClick(Button button) {
        consumers++;
        consumerCount.setText(getResources().getQuantityString(R.plurals.consumers, consumers, consumers));
        sendWorker(CONSUMER);
    }

    @OnClick(R.id.addProducer)
    public void onAddProducerClick(Button button) {
        producers++;
        producerCount.setText(getResources().getQuantityString(R.plurals.producers, producers, producers));
        sendWorker(PRODUCER);
    }

    private void sendWorker(int workerType) {
        if (!isServiceBound) return;
        Message msg;
        switch (workerType) {
            case PRODUCER:
                msg = Message.obtain(null, PRODUCER);
                break;
            case CONSUMER:
                msg = Message.obtain(null, CONSUMER);
                break;
            default:
                throw new IllegalArgumentException("No other option than producer or consumer");
        }
        try {
            producerMessenger.send(msg);
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    private void refresh() {
        adapter.submitList(itemsList);
        listItemsCount.setText(getResources().getQuantityString(R.plurals.listItems, itemsList.size(), itemsList.size()));
        delayAndScroll();
    }

    private void delayAndScroll() {
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                list.getLayoutManager().scrollToPosition(itemsList.size() - 1);
            }
        }, 50);
    }

    private void configureList() {
        LinearLayoutManager layoutManager =
                new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        list.setLayoutManager(layoutManager);
        list.setAdapter(adapter);
    }

    class WorkerBroadcastReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            Bundle bundle = intent.getExtras();
            int workerType;
            if (bundle != null) {
                workerType = bundle.getInt(WORKER_TYPE);
                switch (workerType) {
                    case PRODUCER:
                        itemsList.add(new Item());
                        break;
                    case CONSUMER:
                        if (itemsList.size() > 0) {
                            itemsList.remove(itemsList.size() - 1);
                        } else {
                            String msg = "";
                            if (producers < consumers) {
                                msg += "You have more producers than consumers and t";
                            } else {
                                msg += "T";
                            }
                            msg += "here are no more items to remove, add more producers";
                            Toast.makeText(MainActivity.this, msg, Toast.LENGTH_SHORT).show();
                        }
                        break;
                    default:
                        throw new IllegalArgumentException("No other option than producer or consumer");
                }
                refresh();
            }
        }
    }

}

File: app/src/main/java/net/albertogarrido/studydrivetest/MainActivity.kt
package net.albertogarrido.studydrivetest

import android.content.*
import android.os.*
import android.support.v4.content.LocalBroadcastManager
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*
import net.albertogarrido.studydrivetest.WorkerService.*
import java.util.*

class MainActivity : AppCompatActivity() {

    private var consumers = 0
    private var producers = 0
    private val adapter = ItemsAdapter()
    private val itemsList = ArrayList<Item>(10)

    private var producerMessenger: Messenger? = null
    private val broadcastReceiver: WorkerBroadcastReceiver by lazy { WorkerBroadcastReceiver() }
    private var isServiceBound: Boolean = false

    private val producerConnection = object : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            producerMessenger = Messenger(service)
            isServiceBound = true
        }

        override fun onServiceDisconnected(className: ComponentName) {
            producerMessenger = null
            isServiceBound = false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initButtonListeners()
        configureList()
        refresh()
    }

    override fun onStart() {
        super.onStart()
        bindService(WorkerService.createIntent(this), producerConnection, Context.BIND_AUTO_CREATE)
    }

    public override fun onResume() {
        super.onResume()
        val intentFilter = IntentFilter(WORKER)
        LocalBroadcastManager.getInstance(this).registerReceiver(broadcastReceiver, intentFilter)
    }

    public override fun onPause() {
        super.onPause()
        LocalBroadcastManager.getInstance(this).unregisterReceiver(broadcastReceiver)
    }

    private fun initButtonListeners() {
        addConsumerButton.setOnClickListener {
            consumers++
            consumerCount.text = resources.getQuantityString(R.plurals.consumers, consumers, consumers)
            sendWorker(CONSUMER)
        }
        addProducerButton.setOnClickListener {
            producers++
            producerCount.text = resources.getQuantityString(R.plurals.producers, producers, producers)
            sendWorker(PRODUCER)
        }
    }

    private fun sendWorker(workerType: Int) {
        if (!isServiceBound) return
        val msg: Message = when (workerType) {
            PRODUCER -> Message.obtain(null, PRODUCER)
            CONSUMER -> Message.obtain(null, CONSUMER)
            else -> throw IllegalArgumentException("No other option than producer or consumer")
        }
        try {
            producerMessenger!!.send(msg)
        } catch (e: RemoteException) {
            e.printStackTrace()
        }
    }

    private fun refresh() {
        adapter.submitList(itemsList)
        listItemsCount.text = resources.getQuantityString(R.plurals.listItems, itemsList.size, itemsList.size)
        delayAndScroll()
    }

    private fun delayAndScroll() {
        Handler().postDelayed({ list.layoutManager.scrollToPosition(itemsList.size - 1) }, 50)
    }

    private fun configureList() {
        val layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        list.layoutManager = layoutManager
        list.adapter = adapter
    }

    internal inner class WorkerBroadcastReceiver : BroadcastReceiver() {

        override fun onReceive(context: Context, intent: Intent) {
            val extras = intent.extras

            extras?.let { bundle ->
                val workerType = bundle.getInt(WORKER_TYPE)
                when (workerType) {
                    PRODUCER -> itemsList.add(Item.create())
                    CONSUMER -> {
                        if (itemsList.size > 0) {
                            itemsList.removeAt(itemsList.size - 1)
                        } else {
                            val msg = if (producers < consumers) {
                                "You have more producers than consumers and there are no more items to remove, add more producers"
                            } else {
                                "There are no more items to remove, add more producers"
                            }
                            Toast.makeText(this@MainActivity, msg, Toast.LENGTH_SHORT).show()
                        }
                    }
                    else -> throw IllegalArgumentException("No other option than producer or consumer")
                }
                refresh()
            }
        }
    }

}
